### PR TITLE
Add built-in occurrences filter extension

### PR DESCRIPTION
**Pull request by portertech**
Closes https://github.com/sensu/sensu-extensions/issues/10

### CODE REVIEW

**Comment by agoddard** [comment]
<3

File: spec/filters/occurrences_spec.rb [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        event[:check][:occurrences] = 3
+        @extension.safe_run(event) do |output, status|
+          expect(status).to eq(0)
+          event[:occurrences] = 3
+          @extension.safe_run(event) do |output, status|
+            expect(status).to eq(1)
+            event[:occurrences] = 9
+            @extension.safe_run(event) do |output, status|
+              expect(status).to eq(0)
+              event[:check][:refresh] = 12
```
Comment (cwjohnston) [inline spec/filters/occurrences_spec.rb]: My understanding is that this is testing an event should not match the filter with refresh of 12, do we need a positive match case for refresh as well?

**Comment by fessyfoo** [comment]
I'm curious if  this filter intended to solve the problems that the sensu-plugins occurrence  filter had:
- https://github.com/sensu/sensu/issues/1067#issuecomment-225313629
- https://github.com/sensu/sensu/issues/1092

As it stands now I think it replicates the same issue.  I assume that's expected? 

If it's not expected,  consider this sequence of events: 

`event.occurrences` >`check.occurrences` event not filtered,   handler handles create:
```
{
  "check": {
    "occurrences": 2,
    "status": 2,
    "history": [ 0, 0, 2, 2, 2]
  },
  "occurrences": 3,
  "action": "create"
}
```

next the status changes state from `crit` to `warn` and   `event.occurrences` [is reset to 1](https://github.com/sensu/sensu/blob/69a5b96/lib/sensu/server/process.rb#L487). The event  is filtered.  maybe that's ok. 
```
{
  "check": {
    "occurrences": 2,
    "status": 1,
    "history": [ 0, 2, 2, 2, 1]
  },
  "occurrences": 1,
  "action": "create"
}
```

next comes the resolution event.   status is 0,  and has stored event so `event.occurrences` [has not incremented](https://github.com/sensu/sensu/blob/69a5b96/lib/sensu/server/process.rb#L490-L492), and it's still below the `check.occurrences` threshold so it is filtered leading to confusing situation where on call personnel may have received create notification but not resolve notification. That's at least a significant part of the confusion in the above mentioned bugs. 
 
```
{
  "check": {
    "occurrences": 2,
    "status": 0,
    "history": [ 2, 2, 2, 1, 0]
  },
  "occurrences": 1,
  "action": "resolve"
}
```

Hope that helps make the subtle issue clear to anyone who didn't see that bit already. 

related comment: https://github.com/sensu-plugins/sensu-plugin/issues/134#issuecomment-233123190

> I'm curious why the proposed built in occurrences filter or even regular sensu filters would be in any better position to solve those issues.

**Comment by cwjohnston** [comment]
I've replied to the above question in https://github.com/sensu-plugins/sensu-plugin/issues/134#issuecomment-238608045

**Comment by fessyfoo** [comment]
noting that as of  https://github.com/sensu/sensu-extensions/pull/12 built in extensions have moved out of this repo to individual repos under the [sensu-extensions github project](https://github.com/sensu-extensions).

in particular the occurrences filter has moved to https://github.com/sensu-extensions/sensu-extensions-occurrences  and now includes [a fix for the filtered resolves issue](https://github.com/sensu-extensions/sensu-extensions-occurrences/pull/2) described above.

**Pull request merged**